### PR TITLE
Clarify shared file handler behavior of File::try_clone.

### PR DESCRIPTION
**Pull request by frewsxcv**
Fixes https://github.com/rust-lang/rust/issues/46578.

### CODE REVIEW

File: src/libstd/fs.rs [hunk 1]
```diff
@@ -482,23 +482,45 @@ impl File {
...
     }
 
-    /// Creates a new independently owned handle to the underlying file.
+    /// Create a new `File` that shares the existing underlying file handle.
     ///
-    /// The returned `File` is a reference to the same state that this object
-    /// references. Both handles will read and write with the same cursor
-    /// position.
+    /// Modifying one `File` (e.g. reading, writing, seeking) will result in the
+    /// underlying file handle of the other `File` also adjusting.
```
Comment (frewsxcv) [inline src/libstd/fs.rs]: not a big fan of my wording here or in the previous section. lemme know if y'all have better suggestions

Comment (abonander) [reply src/libstd/fs.rs]: Maybe something along the lines of:

> This means that both `File` instance shares the same cursor; reads, writes and seeks will affect both instances simultaneously.

**Comment by abonander** [comment]
I just realized this could be a bad footgun for someone using a cloned handle with `BufReader<File>`; the buffer won't know to flush and re-read after a seek or modification. Should we warn the reader about this?

File: src/libstd/fs.rs (hunk 1 above)
Comment (frewsxcv) [reply src/libstd/fs.rs]: Thanks for the suggestion! Latest force push incorporates some of that wording

**Comment by frewsxcv** [comment]
> I just realized this could be a bad footgun for someone using a cloned handle with BufReader<File>; the buffer won't know to flush and re-read after a seek or modification. Should we warn the reader about this?

@abonander I think that might be a good idea. Any suggestions on how to word this?

**Comment by abonander** [comment]
Something like

### Note: `BufReader` Data Corruption

Using a cloned handle with `BufReader` could cause data corruption when the other handle is read from, seeked or written to, as the buffer won't know that the data it contains is stale.

--

However, I think this merits some discussion as `BufReader` could instead remember the cursor position from its previous read and check it against the current cursor position before yielding data out of the buffer. However, this would mean a syscall on every `fill_buf()` invocation (?). (Also if the other handle is read or written and then seeked back to the original position then the buffer wouldn't know the difference anyway.)

**Comment by abonander** [comment]
We already have a warning about a similar footgun with `BufReader.get_mut()` so I don't think this requires special treatment.

**Comment by frewsxcv** [comment]
Okay so it sounds like we might want to open a separate issue for that then? What do you think?

**Comment by aidanhs** [comment]
@bors r+ rollup

This looks good to me, and it sounds like the bufreader thing may need more discussion.

**Pull request merged**